prov/shm/src/smr_ep.c

```diff
@@ -1697,16 +1305,25 @@ static int smr_ep_ctrl(struct fid *fid, int command, void *arg)
 			domain = container_of(ep->util_ep.domain,
 					      struct smr_domain,
 					      util_domain.domain_fid);
-			ret = smr_ep_srx_context(domain, ep->rx_size,
```

Comment:
would u mind explaining what's the use case for `ep->srx` is NULL and not? I still could not understand it after reading example SRX setup in the man page.

Reply:
Yeah the man pages are pretty confusing so I don't blame you. This is my flow that I've implemented in rxm for a shm offload:
1. rxm allocates it's own srx internally using the new util_ep_srx_context (util_srx)
2. rxm opens all of the shm resources (fabric/domain, ep, etc)
3. rxm "imports" its srx into shm with fi_srx_context (passing in FI_PEER and the peer context)
     - Since the fi_srx_context is associated with a domain, not an EP, we have to call fi_ep_bind to bind the srx (imported from rxm) to that specific EP. That's where ep->srx gets set. This indicates that we are using a pre-set up srx (whether that was created in fi_srx_context or imported in fi_srx_context. If there is no srx associated with the EP, we have to create a new one (util_ep_srx_context). If there is already an srx associated with it, we can use the owner ops from the imported one, but we still have to allocate a new one to save our peer ops in order to pass it back to the owner

So after the whole set up, rxm and shm will each have their own srx objects
rxm - srx resources (queue, pools, etc); owner_ops = rxm_srx_ops; peer_ops = rxm_start/discard ops;
shm - no srx resources; owner_ops = rxm_srx_ops (imported); peer_ops = shm_start/discard ops

So when shm calls owner_ops.get_msg(), it passes in the shm_srx so that rxm knows what peer ops to use to start that message once it gets a match
When rxm calls owner_ops.get_msg(), it passes in the rxm_srx to use its own peer_ops

Let me know if still still doesn't make sense.

Reply:
> Since the fi_srx_context is associated with a domain, not an EP, we have to call fi_ep_bind to bind the srx (imported from rxm) to that specific EP. That's where ep->srx gets set. This indicates that we are using a pre-set up srx (whether that was created in fi_srx_context or imported in fi_srx_context. If there is no srx associated with the EP, we have to create a new one (util_ep_srx_context). If there is already an srx associated with it, we can use the owner ops from the imported one, but we still have to allocate a new one to save our peer ops in order to pass it back to the owner

So rxm needs to 
1. call `util_srx_context(rxm_domain->util_domain, ... &rxm_ep->srx )` to allocate a new srx, (modify the owner_ops if necessary?)
2. create shm resources via `fi_domain`, `fi_endpoint` ....
3. construct a `fi_peer_srx_context` object `context`, set `context.srx = container_of(rxm_ep->srx, struct fid_peer_srx, ep_fid)`. construct an `attr |= FI_PEER`
4. call `fi_srx_context(&shm_domain, attr, ..., context)` to export the srx to shm domain
5. call `fi_ep_bind(shm_ep, srx, ..)` to bind the srx to shm ep

Inside shm provider, 
1. `smr_ep_bind` will set `smr_ep->srx` as the srx imported by rxm via the fi_ep_bind call
2. `smr_ep_ctrl` will see `smr_ep->srx` already associated, so it will calloc a new srx, get the owner_ops from the imported one, then re-assign the peer_ops as shm's peer_ops.

If this is correct, I am a bit confused on how step 4 is implemented in shm. I can see `smr_srx_context` only set `smr_domain->srx = ((struct fi_peer_srx_context *) (context))->srx;`  when  the attr has `FI_PEER` flag. Then how `smr_domain->srx` is used? It seems like shm only interacts with the peer_srx via ep but not domain.

> So after the whole set up, rxm and shm will each have their own srx objects
rxm - srx resources (queue, pools, etc); owner_ops = rxm_srx_ops; peer_ops = rxm_start/discard ops;
shm - no srx resources; owner_ops = rxm_srx_ops (imported); peer_ops = shm_start/discard ops
So when shm calls owner_ops.get_msg(), it passes in the shm_srx so that rxm knows what peer ops to use to start that message once it gets a match
When rxm calls owner_ops.get_msg(), it passes in the rxm_srx to use its own peer_ops

If rxm's srx has peer_ops as rxm_start/discard ops, how it can make shm start progressing the msg when rxm finds a match in the unexp queue, and the msg was queued by shm? I was referring to these steps in the man page

```
....
5. The application calls fi_recv() / fi_trecv() on owner, posting the
   matching receive buffer.
6. The owner matches the receive with the queued message on the peer.
7. The owner removes the queued request, fills in the rest of the known fields
   and calls the peer->start_msg() / start_tag() function.
```
Shouldn't rxm call shm_start_msg() in this situation? Do I understand the flow incorrectly?

Reply:
In regards to the shm context, the srx is directly set/imported from an owner when FI_PEER is set. If it is not set, we create a new one in shm:
```
int smr_srx_context(struct fid_domain *domain, struct fi_rx_attr *attr,
		    struct fid_ep **rx_ep, void *context)
{
	struct smr_domain *smr_domain;

	smr_domain = container_of(domain, struct smr_domain, util_domain.domain_fid);

	if (attr->op_flags & FI_PEER) {
		smr_domain->srx = ((struct fi_peer_srx_context *) (context))->srx;
		return FI_SUCCESS;
	}
	return util_ep_srx_context(&smr_domain->util_domain, attr->size,
				   SMR_IOV_LIMIT, SMR_INJECT_SIZE, rx_ep);
}
```
The second case is to support the existing fi_srx_context() API where an application may want to share an fi_srx_context between EPs under the same domain. I haven't tested that path, so there definitely could be issues in using that. I'm not sure. But in the context of importing it from the peer, it is still under the domain as a resource but can only be used directly by any EPs under that domain. So the domain doesn't necessarily become relevant in the operations, but rather is the owner of the resource itself. 
So there are two cases that we are interested in for this support - either an fi_srx_context was imported from an owner with FI_PEER to use shm as an intranode offload provider or there is no peer in which case shm creates its own srx when it does not detect an existing srx (this happens on enable). In the other, untested case where the application is sharing a shm srx to use among shm EPs, the individual EP would not create a second one on enable, but rather set the owner ops to use.


For the second question about the flow, and which set of start/discard ops to use, this is the reason why we need two srxs, one for each peer. In the flow, when a peer calls get_msg(), it passes in the srx to use for processing the message. So shm would pass in its srx (which has the shm_start_ops) and rxm would pass in its srx (which has the rxm_start_ops). The owner sets the srx to use for processing the message later.

When rxm saves an unexpected message for later, it allocates a new rx_entry and sets the peer_entry.srx to the inputted srx (if shm called it, it would be the shm srx with the shm start ops and if rxm called it, it would be the rxm start ops):
```
	match_attr.addr = addr;
	dlist_entry = dlist_find_first_match(&srx_ctx->recv_queue.list,
					     srx_ctx->recv_queue.match_func,
					     &match_attr);
	if (!dlist_entry) {
		util_entry = util_alloc_rx_entry(srx_ctx);
		if (!util_entry) {
			ret = -FI_ENOMEM;
		} else {
			util_entry->peer_entry.owner_context = NULL;
			util_entry->peer_entry.addr = addr;
			util_entry->peer_entry.size = size;
			util_entry->peer_entry.srx = srx;
			*rx_entry = &util_entry->peer_entry;
			ret = -FI_ENOENT;
		}
		goto out;
	}
```
When a new receive gets posted, it searches the unexpected queue to see if there is a waiting unexpected message that came in that would match that receive and calls the start_msg for the peer srx that added that unexpected message:
```
	ofi_spin_lock(&srx->lock);
	dlist_entry = dlist_remove_first_match(&unexp_queue->list,
					       unexp_queue->match_func,
					       &match_attr);
	if (!dlist_entry) {
		rx_entry = util_get_recv_entry(srx, iov, desc, iov_count, addr,
					       context, tag, ignore, flags);
		if (!rx_entry)
			ret = -FI_ENOMEM;
		else
			dlist_insert_tail((struct dlist_entry *)
					  (&rx_entry->peer_entry),
					  &recv_queue->list);
		ofi_spin_unlock(&srx->lock);
		return ret;
	}
	ofi_spin_unlock(&srx->lock);

	rx_entry = container_of(dlist_entry, struct util_rx_entry, peer_entry);
	util_init_rx_entry(srx, rx_entry, iov, desc, iov_count, addr, context,
			   tag, flags);

	return srx->peer_srx.peer_ops->start_msg(&rx_entry->peer_entry);
```

Reply:
Thank you very much for the explanation!

>When a new receive gets posted, it searches the unexpected queue to see if there is a waiting unexpected message that came in that would match that receive and calls the start_msg for the peer srx that added that unexpected message:

Yes, I get all of what you are talking for the expected flow. When peers call the `get_msg` op, it will set the srx to be used in `rx_entry->peer_entry.srx`, and when owner processes the matched rx entry from the unexp queue, it should find which srx to use from the rx entry.  

But looking at the code in `util_srx_generic_recv`,  the `srx` in `srx->peer_srx.peer_ops->start_msg(&rx_entry->peer_entry);`  is derived from the local ep, 

```
srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
```
 I was expecting it should be derived from the peer_srx in the matched `rx_entry`? The `rx_entry->peer_entry` passed to the op does not really make it toggle which srx to use.